### PR TITLE
Fix time_bucket DST handling with timezone and offset

### LINKED ISSUE #7059: [Bug]: time_bucket function doesn't account for time zone when applying offset
### What type of bug is this?

Incorrect result

### What subsystems and features are affected?

Other

### What happened?

# Description
I'm using the `time_bucket` function with both `timezone` and `offset` arguments, and I expected the offset to be applied _after_ the time zone so that my 15 minute offset jumps ahead _an hour and 15 minutes_ when DST occurs.

To clarify, in the screenshot below the "local_time" is in the `Asia/Amman` time zone, and when adding a 15 minute interval to "2021-03-25 23:45", the result should be "2021-03-26 01:00" instead of "2021-03-26 00:00" (an hour was skipped because of DST)
image

The `time_bucket` function doesn't take that into consideration because when running the below query I'd expect it to sum up the first 4 rows (all of March 25th, and the first row of March 26th because it's been offset 15 minutes).
```sql
SELECT
    time_bucket('1 day'::interval, time,
        "timezone" := 'Asia/Amman',
        "offset" := '15 minutes'::interval
    ) as day_bucket,
    metric_id,
    sum(value) as sum,
    string_agg(value::text, ', ') as grouped_values
FROM metric_data
GROUP BY day_bucket, metric_id;
```
But instead the actual output is this:
image

# Workaround
A workaround for this issue would be to actually not use the offset parameter on the time_bucket function and instead apply the offset directly on your timestamp column since that column is (hopefully) in UTC:

```sql
SELECT
    time_bucket(
        '1 day'::interval,
        time - '15 minutes'::interval,
        "timezone" := 'Asia/Amman',
    ) as day_bucket,
    metric_id,
    sum(value) as sum,
    string_agg(value::text, ', ') as grouped_values
FROM metric_data
GROUP BY day_bucket, metric_id;
```
output:
image

### TimescaleDB version affected

2.5.1

### PostgreSQL version used

14.2

### What operating system did you use?

Windows 10 x64

### What installation method did you use?

Docker

### What platform did you run on?

On prem/Self-hosted

### Relevant log output and stack trace

_No response_

### How can we reproduce the bug?

```bash
SQL Setup

CREATE TABLE metric_data (
    time timestamp with time zone,
    metric_id varchar(20),
    value numeric
);

INSERT INTO metric_data(time, metric_id, value)
VALUES ('2021-03-25 21:15:00+00:00', 'm1', 2),
       ('2021-03-25 21:30:00+00:00', 'm1', 6),
       ('2021-03-25 21:45:00+00:00', 'm1', 15),
       ('2021-03-25 22:00:00+00:00', 'm1', 5),
       ('2021-03-25 22:15:00+00:00', 'm1', 1),
       ('2021-03-25 22:30:00+00:00', 'm1', 9)
;


The output of this query is not accurate
```sql
SELECT
    time_bucket(
            "bucket_width" := '1 day'::interval,
            "ts" := time,
            "timezone" := 'Asia/Amman',
            "offset" := '15 minutes'::interval
    ) as day_bucket,
    metric_id,
    sum(value) as sum
FROM metric_data
GROUP BY day_bucket, metric_id;
```
```

### LINKED ISSUE #8851: [Bug]: Wrong output of time_bucket() when using timezone parameter during DST switch
### What type of bug is this?

Incorrect result

### What subsystems and features are affected?

Query executor

### What happened?

The following query returns the result shown below:

```SQL
SELECT
    time_bucket('30 seconds', "timestamp", timezone => 'Europe/Berlin', "offset" => '-15 seconds'::interval) + '15 seconds'::interval AS "bucket", 
    AVG("value") AS "value"
FROM "measurement"
WHERE 
    "sensor_id"=1589 AND 
    "timestamp">='2025-10-26 01:59:00+02:00' AND "timestamp"<'2025-10-26 03:01:00+01:00' 
GROUP BY "sensor_id", "bucket"
ORDER BY "bucket";
```

Shortened output:

```
         bucket         |       value        
------------------------+--------------------
 2025-10-26 01:59:00+02 |   573.555419921875
 2025-10-26 01:59:30+02 |  579.8365478515625
 2025-10-26 02:00:00+02 |  579.4089965820312
 2025-10-26 02:00:30+01 |  578.9915771484375
 2025-10-26 02:01:00+01 |  578.7264404296875
 2025-10-26 02:01:30+01 |  578.8228759765625
 2025-10-26 02:02:00+01 |   578.565185546875
[...]
 2025-10-26 02:59:00+01 |  472.4388427734375
 2025-10-26 02:59:30+01 |  472.2878112792969
 2025-10-26 02:00:00+02 |  472.1482238769531
 2025-10-26 02:00:30+01 |   471.944091796875
 2025-10-26 02:01:00+01 |  471.6407775878906
 2025-10-26 02:01:30+01 | 470.87664794921875
 2025-10-26 02:02:00+01 |  469.7684020996094
 2025-10-26 02:02:30+01 |     468.1826171875
[...]
 2025-10-26 02:58:30+01 |   570.677490234375
 2025-10-26 02:59:00+01 |    570.53466796875
 2025-10-26 02:59:30+01 |  570.5411987304688
 2025-10-26 03:00:00+01 |        570.4609375
 2025-10-26 03:00:30+01 |  570.4595336914062

```

**This result is incorrect** (see below for expected result)

When skipping the `timezone` parameter in the `time_bucket()` call like so:

```SQL
SELECT 
    time_bucket('30 seconds', "timestamp", "offset" => '-15 seconds'::interval) + '15 seconds'::interval AS "bucket", 
    AVG("value") AS "value"
FROM "measurement"
WHERE 
    "sensor_id"=1589 AND 
    "timestamp">='2025-10-26 01:59:00+02:00' AND "timestamp"<'2025-10-26 03:01:00+01:00' 
GROUP BY "sensor_id", "bucket"
ORDER BY "bucket";
```

... I get the expected (again shortened) result:

```
         bucket         |       value        
------------------------+--------------------
 2025-10-26 01:59:00+02 |   573.555419921875
 2025-10-26 01:59:30+02 |  579.8365478515625
 2025-10-26 02:00:00+02 |  579.4089965820312
 2025-10-26 02:00:30+02 |  578.9915771484375
 2025-10-26 02:01:00+02 |  578.7264404296875
 2025-10-26 02:01:30+02 |  578.8228759765625
 2025-10-26 02:02:00+02 |   578.565185546875
[...]
 2025-10-26 02:59:00+02 |  472.4388427734375
 2025-10-26 02:59:30+02 |  472.2878112792969
 2025-10-26 02:00:00+01 |  472.1482238769531
 2025-10-26 02:00:30+01 |   471.944091796875
 2025-10-26 02:01:00+01 |  471.6407775878906
 2025-10-26 02:01:30+01 | 470.87664794921875
 2025-10-26 02:02:00+01 |  469.7684020996094
 2025-10-26 02:02:30+01 |     468.1826171875
[...]
 2025-10-26 02:58:30+01 |   570.677490234375
 2025-10-26 02:59:00+01 |    570.53466796875
 2025-10-26 02:59:30+01 |  570.5411987304688
 2025-10-26 03:00:00+01 |        570.4609375

```

Is is expected but not observed that:
- there is only one switch from +02 to +01, not the other way around, not more than once, not even for a single row
- the switch from +02 to +01 should happen between 02:59:30+02 (in this case) and the following 02:00:00+01

Not as well: The values are the same between the two query results, which is even weirder. 

I can not just get rid of the `timezone` parameter because it is already a workaround for another DST related issue :cry:

Removing the `offset` parameter as well as the `+'15 seconds'::interval` part results in a quite similarly wrong output, except for both 02:00:00 timestamps are in timezone +01.

### TimescaleDB version affected

2.22.1

### PostgreSQL version used

16.10

### What operating system did you use?

AlmaLinux 9 x86_64

### What installation method did you use?

RPM

### What platform did you run on?

On prem/Self-hosted

### Relevant log output and stack trace

```bash

```

### How can we reproduce the bug?

```bash
1. Setup a simple hypertable:


CREATE TABLE "measurement" (
  "timestamp" timestamptz NOT NULL DEFAULT CURRENT_TIMESTAMP,
  "sensor_id" integer NOT NULL,
  "value" real NOT NULL,
  PRIMARY KEY ("timestamp", "sensor_id")
); 


2. Fill it with data (at least two values per hour should be sufficient) around a DST switch event (e.g. 2025-10-26 01:00:00+02 to 2025-10-26 03:00:00+01)
3. Execute the queries above and be amazed
```

### LINKED ISSUE #9136: [Bug]: Timestamp assigned to a time bucket that starts after the timestamp
### What type of bug is this?

Incorrect result

### What subsystems and features are affected?

Other

### What happened?

The time_bucket() function can sometimes produce buckets where the start of the bucket is _after_ the timestamp that was bucketed. This happens when the origin and the timestamp are in different daylight savings time zones (e.g., origin in EST and timestamp in EDT).

### TimescaleDB version affected

2.24.0

### PostgreSQL version used

18.1

### What operating system did you use?

Arch Linux

### What installation method did you use?

Source

### What platform did you run on?

Not applicable

### Relevant log output and stack trace

```bash
SET timezone TO 'America/New_York';
SET
  -- Fall back: November 3, 2024 - at 2:00 AM EDT, clocks go back to 1:00 AM EST
  -- The hour 1:00-2:00 AM local happens TWICE
  SELECT
      label,
      ts,
      ts AT TIME ZONE 'UTC' as ts_utc,
      time_bucket('1 hour', ts, 'America/New_York',
                  '2000-01-01 00:00:00 America/New_York'::timestamptz) as bucket_start,
      time_bucket('1 hour', ts, 'America/New_York',
                  '2000-01-01 00:00:00 America/New_York'::timestamptz) AT TIME ZONE 'UTC' as bucket_utc
  FROM (VALUES
      ('1:30 AM EDT (first)',  '2024-11-03 01:30:00-04'::timestamptz),
      ('1:30 AM EST (second)', '2024-11-03 01:30:00-05'::timestamptz)
  ) AS t(label, ts);
        label         |              ts              |          ts_utc          |         bucket_start         |        bucket_utc
----------------------+------------------------------+--------------------------+------------------------------+--------------------------
 1:30 AM EDT (first)  | Sun Nov 03 01:30:00 2024 EDT | Sun Nov 03 05:30:00 2024 | Sun Nov 03 01:00:00 2024 EST | Sun Nov 03 06:00:00 2024
 1:30 AM EST (second) | Sun Nov 03 01:30:00 2024 EST | Sun Nov 03 06:30:00 2024 | Sun Nov 03 01:00:00 2024 EST | Sun Nov 03 06:00:00 2024
(2 rows)

SET timezone TO 'America/New_York';
SET
  -- Fall back: November 3, 2024
  -- A timestamp at 1:30 AM EDT (first occurrence) gets assigned to a bucket
  -- that starts AFTER the timestamp!
  SELECT
      ts,
      ts AT TIME ZONE 'UTC' as ts_utc,
      time_bucket('1 hour', ts, 'America/New_York',
                  '2000-01-01 00:00:00 America/New_York'::timestamptz) as bucket_start,
      time_bucket('1 hour', ts, 'America/New_York',
                  '2000-01-01 00:00:00 America/New_York'::timestamptz) AT TIME ZONE 'UTC' as bucket_utc,
      ts < time_bucket('1 hour', ts, 'America/New_York',
                       '2000-01-01 00:00:00 America/New_York'::timestamptz) as timestamp_before_bucket
  FROM (VALUES ('2024-11-03 01:30:00-04'::timestamptz)) AS t(ts);
              ts              |          ts_utc          |         bucket_start         |        bucket_utc        | timestamp_before_bucket
------------------------------+--------------------------+------------------------------+--------------------------+-------------------------
 Sun Nov 03 01:30:00 2024 EDT | Sun Nov 03 05:30:00 2024 | Sun Nov 03 01:00:00 2024 EST | Sun Nov 03 06:00:00 2024 | t
(1 row)
```

### How can we reproduce the bug?

```bash
SET timezone TO 'America/New_York';

  -- Fall back: November 3, 2024
  -- A timestamp at 1:30 AM EDT (first occurrence) gets assigned to a bucket
  -- that starts AFTER the timestamp!

  SELECT
      ts,
      ts AT TIME ZONE 'UTC' as ts_utc,
      time_bucket('1 hour', ts, 'America/New_York',
                  '2000-01-01 00:00:00 America/New_York'::timestamptz) as bucket_start,
      time_bucket('1 hour', ts, 'America/New_York',
                  '2000-01-01 00:00:00 America/New_York'::timestamptz) AT TIME ZONE 'UTC' as bucket_utc,
      ts < time_bucket('1 hour', ts, 'America/New_York',
                       '2000-01-01 00:00:00 America/New_York'::timestamptz) as timestamp_before_bucket
  FROM (VALUES ('2024-11-03 01:30:00-04'::timestamptz)) AS t(ts);
```

### PR DESCRIPTION
**Pull request by tjgreen42**
## Summary
Fix `time_bucket()` with timezone parameter during DST transitions:
- Apply offset in UTC space to avoid creating non-existent local times
- Handle DST fall-back ambiguity when converting bucket back to timestamptz

### #7059: Offset + timezone during DST spring-forward
The offset was applied to local time after timezone conversion, which could create non-existent times.

**Example (Asia/Amman):** On 2021-03-26, clocks jumped from 00:00 to 01:00. Subtracting 15 minutes from 01:00 local produced 00:45, which doesn't exist.

### #8851: Offset + timezone during DST fall-back  
Same root cause as #7059, causing incorrect timestamps around fall-back transitions.

### #9136: Origin + timezone during DST fall-back
When using `origin` parameter during DST fall-back, the bucket could start AFTER the input timestamp. Postgres picks the standard time interpretation for ambiguous local times, but if the input was in daylight time, the bucket ended up in the future.

## Solution
1. Apply offset in UTC space (before timezone conversion) instead of local time
2. After converting bucket back to timestamptz, subtract periods as needed until bucket ≤ timestamp

```
Before: timestamptz → local → subtract offset → bucket → add offset → timestamptz
After:  timestamptz → subtract offset (UTC) → local → bucket → [adjust for DST] → timestamptz → add offset (UTC)
```

## Testing
Added regression tests for DST boundary cases from all three issues.

Fixes #7059
Fixes #8851
Fixes #9136

### CODE REVIEW

File: test/sql/timestamp.sql.in [hunk 1]
```diff
@@ -335,6 +335,13 @@ FROM unnest(ARRAY[
     TIMESTAMP WITH TIME ZONE '2017-11-05 15:05:00+07'
     ]) AS time;
 
+-- GitHub issue #7059: time_bucket with timezone + offset across DST boundary
+-- Asia/Amman skipped 00:00-00:59 on 2021-03-26 (clocks jump from 00:00 to 01:00)
+-- The offset must be applied in UTC space to avoid creating non-existent local times
+SET timezone TO 'UTC';
```
Comment (akuzm) [inline test/sql/timestamp.sql.in]: The test itself doesn't depend on the current tz, right? Let's not change it, for clarity.

File: test/sql/timestamp.sql.in [hunk 2]
```diff
@@ -335,6 +335,13 @@ FROM unnest(ARRAY[
     TIMESTAMP WITH TIME ZONE '2017-11-05 15:05:00+07'
     ]) AS time;
 
+-- GitHub issue #7059: time_bucket with timezone + offset across DST boundary
```
Comment (akuzm) [inline test/sql/timestamp.sql.in:338]: Would be good to add a test for the second issue because it describes a slightly different case.

Comment (tjgreen42) [reply test/sql/timestamp.sql.in:338]: Okay can do.  (I tested the other one locally after looking for related issues that this PR might have fixed).

File: test/sql/timestamp.sql.in (hunk 1 above)
Comment (tjgreen42) [reply test/sql/timestamp.sql.in]: Removed

**Comment by akuzm** [comment]
Looks that it also fixes this one https://github.com/timescale/timescaledb/issues/9136 ?

**Comment by erimatnor** [comment]
> Looks that it also fixes this one #9136 ?

Yeah, looks related. We should add the test in that issue to this PR. @tjgreen42

**Comment by tjgreen42** [comment]
> > Looks that it also fixes this one #9136 ?
> 
> Yeah, looks related. We should add the test in that issue to this PR. @tjgreen42

Slightly different case actually -- this PR fixes the situation for `timestamp` and `offset`, but not for `timestamp` and `origin`.  But let me see about folding in a fix for `origin` too.

File: .unreleased/pr_9129 [hunk 3]
```diff
@@ -0,0 +1 @@
+Fixes: #7059 #8851 #9136 Fix time_bucket with timezone during DST
```
Comment (svenklemm) [inline .unreleased/pr_9129]: This is supposed to be the PR number since we compile changelog from this file, the actual bugs should go into commit message/linked to PR. Any external contributors reporting any of the bugs should get a thank you line

Comment (svenklemm) [inline .unreleased/pr_9129]:
```suggestion
Fixes: #9129 Fix time_bucket with timezone during DST
Thanks: @cracksalad and @eyadmba for reporting a bug with timezone handling in time_bucket
```

**Review by svenklemm — APPROVED** [review]

File: .unreleased/pr_9129 (hunk 3 above)
Comment (tjgreen42) [reply .unreleased/pr_9129]: Got it, thanks
